File: First-Edition/BookSamples/rosetta-stone/rust/rsc1p4_correctness_validation/src/cellular/language/extensions.rs
use crate::automaton::CellConditionWrapper;
use crate::automaton::IState;
use crate::automaton::IStep;
use crate::automaton::INeighborhood;
use crate::automaton::NeighborhoodWrapper;
use crate::automaton::StateWrapper;
use crate::automaton::RuleWrapper;

use std::marker::PhantomData;
use type_level::IInterface;
use tl_list_lib::I32List;
use tl_list_lib::HList;
use tl_str_list::TlStr;


pub struct NeighborsCountImpl<S, Cnts: I32List>
  (PhantomData::<(S, Cnts)>);

pub type NeighborsCount<S, Cnts> =
  CellConditionWrapper<NeighborsCountImpl<S, Cnts>>;


pub struct AdjacentsLvlImpl<const LVL: u8>;

pub type AdjacentsLvl<const LVL: u8> = NeighborhoodWrapper<AdjacentsLvlImpl<LVL>>;


pub struct StateImpl<Name: TlStr, const IDX: u8> (PhantomData::<Name>);

pub type State<Name, const IDX: u8> = StateWrapper<StateImpl<Name, IDX>>;


pub struct RuleImpl <
    StDict: HList<IState>,
    Name: TlStr,
    Code: TlStr,
    Neighborhood: IInterface<INeighborhood>,
    Step: IInterface<IStep>>
      (PhantomData::<(StDict, Name, Code, Neighborhood, Step)>);

pub type Rule<StDict, N, C, Nh, S>
  = RuleWrapper<RuleImpl<StDict, N, C, Nh, S>>;

File: First-Edition/BookSamples/rosetta-stone/scala3/rsc2p3_type_level_interfaces/src/main/scala/cellular/Extensions.scala
package cellular
package extensions

import cellular.typelevel._
import cellular.language._


// Parameterized implementation type, option 1: associated types
sealed trait StateImpl:
  type Name <: String
  type Index <: Int

type State[N <: String, I <: Int] =
  IStateObj.MkState[StateImpl { type Name = N; type Index = I }]


// Parameterized implementation type, option 2: type parameters
case class NeighborsCountImpl[
  S <: IState,
  Cnts <: IntList]()

type NeighborsCount[S <: IState, Cnts <: IntList] =
  ICellConditionObj.MkCellCondition[
    NeighborsCountImpl[S, Cnts]]


case class StateTransitionImpl[
  From <: IState,
  To <: IState,
  Conds <: ICellCondition]()

type StateTransition[
  From <: IState,
  To <: IState,
  Conds <: ICellCondition] =
  IStateTransitionObj.MkStateTransition[
    StateTransitionImpl[From, To, Conds]]


case class AdjacentsLvlImpl[
  Lvl <: Int & Singleton]()

type AdjacentsLvl[Lvl <: Int & Singleton] =
  INeighborhoodObj.MkNeighborhood[
    AdjacentsLvlImpl[Lvl]]


case class StepImpl[
  DefState <: IState,
  Transitions <: HList[IStateTransition]]()

type Step[
  DefState <: IState,
  Transitions <: HList[IStateTransition]] =
  IStepObj.MkStep[
    StepImpl[DefState, Transitions]]



case class RuleImpl[
  Name <: String & Singleton,
  Code <: String & Singleton,
  Neighborhood <: INeighborhood,
  Step <: IStep]()

type Rule[
  Name <: String & Singleton,
  Code <: String & Singleton,
  Neighborhood <: INeighborhood,
  Step <: IStep] =
  IRuleObj.MkRule[
    RuleImpl[Name, Code, Neighborhood, Step]]
